compiler/rustc_builtin_macros/src/format.rs

```diff
@@ -369,7 +377,7 @@ impl<'a, 'b> Context<'a, 'b> {
 
         let count = self.pieces.len()
             + self.arg_with_formatting.iter().filter(|fmt| fmt.precision_span.is_some()).count();
-        if self.names.is_empty() && !numbered_position_args && count != self.args.len() {
+        if self.names.is_empty() && !numbered_position_args && count != self.num_args() {
```

Comment:
Normally this (positive) branch would result in a message `1 positional argument in format string, but no arguments were given`, whereas now adding an implicit capture gets `invalid reference to positional argument 0 (no arguments were given)`. Ideally this wording would be kept, even with implicit args; I _think_ the proper check would then be `self.names.len() == self.num_captured_args && !numbered_position_args && count != self.num_args()`?

Reply:
Then the `let count =` above would also have to be adjusted to exclude all placeholders that refer to captures, which is a bit more complicated.

I agree the diagnostics here can be improved when there's captured arguments involved, but I rather leave that to a separate change that can fix all those cases properly.